server/src/main/java/org/elasticsearch/synonyms/SynonymsManagementAPIService.java

```diff
@@ -0,0 +1,164 @@
+
+import static org.elasticsearch.index.mapper.MapperService.SINGLE_MAPPING_NAME;
+import static org.elasticsearch.xcontent.XContentFactory.jsonBuilder;
+
+public class SynonymsManagementAPIService {
+    public static final String SYNONYMS_INDEX = ".synonyms";
+    public static final String SYNONYMS_ORIGIN = "synonyms";
+
+    public static final String SYNONYMS_FEATURE_NAME = "synonyms";
+    public static final String SYNONYM_SET_FIELD = "synonym_set";
+    public static final String SYNONYM_FIELD = "synonym";
+
+    private final Client client;
+
+    public static final SystemIndexDescriptor SYNONYMS_DESCRIPTOR = SystemIndexDescriptor.builder()
+        .setIndexPattern(SYNONYMS_INDEX + "*")
+        .setDescription("Synonyms index for synonyms managed through APIs")
+        .setPrimaryIndex(SYNONYMS_INDEX)
+        .setMappings(mappings())
+        .setSettings(settings())
+        .setVersionMetaKey("version")
+        .setOrigin(SYNONYMS_ORIGIN)
+        .build();
+
+    public SynonymsManagementAPIService(Client client) {
+        // TODO Should we set an OriginSettingClient? We would need to check the origin at AuthorizationUtils if we set an
+        this.client = new OriginSettingClient(client, SYNONYMS_ORIGIN);
+    }
+
+    private static XContentBuilder mappings() {
+        try {
+            XContentBuilder builder = jsonBuilder();
+            builder.startObject();
+            {
+                builder.startObject(SINGLE_MAPPING_NAME);
+                {
+                    builder.startObject("_meta");
+                    {
+                        builder.field("version", Version.CURRENT.toString());
+                    }
+                    builder.endObject();
+                    builder.field("dynamic", "strict");
+                    builder.startObject("properties");
+                    {
+                        builder.startObject("synonym");
+                        {
+                            builder.field("type", "text");
+                        }
+                        builder.endObject();
+                        builder.startObject(SYNONYM_SET_FIELD);
+                        {
+                            builder.field("type", "keyword");
+                        }
+                        builder.endObject();
+                    }
+                    builder.endObject();
+                }
+                builder.endObject();
+            }
+            builder.endObject();
+            return builder;
+        } catch (IOException e) {
+            throw new UncheckedIOException("Failed to build mappings for " + SYNONYMS_INDEX, e);
+        }
+    }
+
+    public void putSynonymSet(String resourceName, SynonymSet synonymSet, ActionListener<PutSynonymsAction.Response> listener) {
+
+        // TODO Add synonym rules validation
+
+        // Delete synonym set if it existed previously
+        DeleteByQueryRequest dbqRequest = new DeleteByQueryRequest(SYNONYMS_INDEX).setQuery(
```

Comment:
PUT request first deletes all synonym docs using a delete by query, and then uses a bulk insert

Reply:
What's the reason for using `delete by query` request?  Why do we need to all documents belonging to the this synonymSetID on update? I thought the whole point of using multiple documents is that we don't need to do this. 

I think what we should do instead is to do a bulk update request:
- create SynonymRules that don't have IDs
- update SynonymRules that have IDs

Reply:
The PUT update will replace the whole synonym set, so it makes sense to me to first remove all existing docs. Keep in mind that the IDs might not have been included from the user, and I think it's good to let the user change the whole synonym set without needing to include the rules IDs.

> I thought the whole point of using multiple documents per synonymSetID is to avoid this, and work on individual document base.

Correct! But that would be done at the individual synonym set level. So it will be possible to do:

`PUT /_synonyms/<synonym_set>/<rule_id> `

to override a single rule. But using

`PUT /_synonyms/<synonym_set>`

will replace the whole synonym set, effectively removing all synonym set rules and adding them from the request.

Does it make sense?